dask/array/core.py

```diff
@@ -857,28 +857,36 @@ class Array(DaskMethodsMixin):
         Shape of the entire array
     chunks: iterable of tuples
         block sizes along each dimension
+    dtype : str or dtype
+        Typecode or data-type for the new Dask Array
+    meta : empty ndarray
+        empty ndarray created with same NumPy backend, ndim and dtype as the
+        Dask Array being created (overrides dtype)
 
     See Also
     --------
     dask.array.from_array
     """
-    __slots__ = 'dask', '_name', '_cached_keys', '_chunks', 'dtype'
+    __slots__ = 'dask', '_name', '_cached_keys', '_chunks', '_meta'
 
-    def __new__(cls, dask, name, chunks, dtype, shape=None):
+    def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
         self = super(Array, cls).__new__(cls)
         assert isinstance(dask, Mapping)
         if not isinstance(dask, HighLevelGraph):
             dask = HighLevelGraph.from_collections(name, dask, dependencies=())
         self.dask = dask
         self.name = name
-        if dtype is None:
-            raise ValueError("You must specify the dtype of the array")
-        self.dtype = np.dtype(dtype)
+        if dtype is None and meta is None:
+            raise ValueError("You must specify the meta or dtype of the array")
 
-        self._chunks = normalize_chunks(chunks, shape, dtype=self.dtype)
+        self._chunks = normalize_chunks(chunks, shape, dtype=dtype or meta.dtype)
         if self._chunks is None:
             raise ValueError(CHUNKS_NONE_ERROR_MESSAGE)
 
+        if meta is None and dtype is not None:
+            meta = np.empty((0,) * self.ndim, dtype=dtype)
+        self._meta = meta
```

Comment:
meta might be a sizable array here, if so we should probably shrink it down to a single element.  I think that we should have a normalize_meta function.  Actually, I think that in dask dataframe this is called `make_meta`.

It is much more complex though because there we coerce odd inputs, like dicts

https://github.com/dask/dask/blob/9733ab568c732181ec7a786b28944bab4acfd8db/dask/dataframe/utils.py#L272-L304

We should probably use the same name though.

Reply:
`meta` here should be an 0-length array, unless a wrong array is passed, we could add checks for that. Maybe you haven't seen, but we have a `meta_from_array` function in https://github.com/dask/dask/pull/4543/files#diff-3598e11c3b3a16881fecff61abdf1f4cR46. I see your point in having the same name, but I think `meta_from_array` leaves less room for ambiguity, so I'd prefer something that makes a distinction between creating dataframe and array meta. Any objections here?